lib/service/base.rb

```diff
@@ -257,6 +261,39 @@ def run_step
       end
     end
 
+    # @!visibility private
+    class LockStep < Step
+      include StepsHelpers
+
+      attr_reader :steps
+
+      def initialize(*keys, &block)
+        @keys = keys
+        @steps = []
+        instance_exec(&block)
+      end
+
+      def run_step
+        success =
+          DistributedMutex.synchronize(lock_name) do
+            steps.each { |step| step.call(instance, context) }
+            :success
+          end
+
+        if success != :success
+          context[result_key].fail(locked: true)
+          context.fail!
+        end
+      end
+
+      def lock_name
+        [
+          context.__service_class__.to_s.underscore,
+          *@keys.map { |key| context[:params].send(key) }.filter(&:presence),
```

Comment:
I think we can encounter some problems here:
- `context[:params]` can hold objects and not just primitives depending on what happens in the contract
- if we don’t keep the key name, we can be in a situation where two different keys have the same value (can be easy with ids), thus not referring to the proper lock.
- do we need to filter for presence? Depending on the key and the service, differentiating between `"    "` and `nil` can be a legit case. I guess we should trust the values coming out of the contract?

The simplest way of addressing those issues is probably to output the keys followed by their value instead of just the values, wdyt?